xds/src/main/java/io/grpc/xds/XdsClient.java

```diff
@@ -643,5 +650,14 @@ static XdsChannelFactory getInstance() {
      * Creates a channel to one of the provided management servers.
      */
     abstract ManagedChannel createChannel(List<ServerInfo> servers);
+
+    /**
+     * Gets features of the server that the channel is created for. Value is only available
+     * after {@link #createChannel} is called.
+     */
+    @Nullable
+    List<?> getSelectedServerFeatures() {
```

Comment:
It doesn't make sense to put such an API here. This makes a factory class stateful. 

Maybe it is time to eliminate `XdsChannelFactory`, instead we do need something like [`Helper.createResolvingOobChannel()`](https://github.com/grpc/grpc-java/blob/2204beba2e580ec87bc9972bb8f2565877a0af91/api/src/main/java/io/grpc/LoadBalancer.java#L1054), but accessible inside NameResolver.

Last time when I was doing some local e2e test for traffic splitting and routing, the gRPC channel is using plaintext, but the xDS channel is fixed to use TLS (if not google_default) instead of falling back to use the same configuration as its parent channel. This is broken for Java. It is not bitting us because we are always using TLS if not google_default in production.

We should support such a feature correctly. Basically, the creation of xDS channel will be done outside XdsClientImpl. For EDS only's code path, we are roughly able to do it with `Helper.createResolvingOobChannel()`, so not a problem for migrating.

For now, I think a better way is to change the `createChannel` API for `XdsChannelFactory`. You could define a simple class that contains a Channel and server features.

Reply:
The problem is coming from the factory or whatever structure being used for XdsClientIml consumes a list of server_urls. When you create the channel from it, you must know the index of the server_url that is selected, so that you can know the corresponding server feature. If it's only a single server_url, there is no problem at all.

Reply:
A workaround is we pass just one `ServerInfo` instead of `List<ServerInfo>` to `XdsClientImpl` as well as `XdsChannelFactory`. Because we are actually not supporting multiple management server yet, and the current `XdsChannelFactory` has a hack to pick the hard coded first ServerInfo in the list.

Reply:
This is what I mean:

```
class XdsChannel {
  ManagedChannel channel;
  boolean useProtocolV3;  // or something similar if you have a better idea
}

XdsChannel createChannel(List<ServerInfo> servers) {
  ...
}
```

The output of creating channel contains the server feature.

Reply:
I also thought about that, which introduces yet another class and I had thought you wouldn't like it. I'm totally okay with it.

Reply:
:( Why would you think introducing a new class is bad? 😢  This looks fairly natural with channel and protocol feature encapsulated together.

Reply:
Done.